text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+package in ADC regions, the right dependencies will already be present in the
+`node_modules` directory.
+
+#### .jsii Files
+
+There are two files, .jsii.tabl.json and .jsii, bundled in `aws-cdk-lib` for the
+JSII runtime to work for any non-NodeJS languages. And, to enable code and
+documentation generation from the published npm package. Today, they make up 42%
+of the package size. And in the future, after adding more Lambda Layers, the .jsii files would
+compose 30% of the package size.
+
+We will compress the .jsii.tabl.json and .jsii files. Compressing with gzip
+currently creates a 4 MB and 4.7 MB file, respectively. This change will impact
+the JSII runtime, and Construct Hub. NodeJS JSII runtimes don't need to use
+either file, and non-NodeJS runtimes would decompress the .jsii file when
+needed. Construct Hub generates documentation from the .jsii.tabl.json file, and
+it will also decompress this file as needed.
+
+#### Source Map Files (.js.map)
+
+These files currently makeup 22% of the total size of `aws-cdk-lib`. Source map
+files map from the transformed source to the original source. In the case of
+`aws-cdk-lib`, these are not actually useful today, since we do not ship the
+original `.ts` files. We will remove the source maps from the released package.
+
+#### Javascript Files (.js)
+
+The Javascript files make up about 10% of the overall package size. Today, they
+are already minified in the released package. There is not much else that can be
+done to make them smaller.
+
+#### Type Declaration Files (.d.ts)
+
+We can minify these files, which could result in a space saving of up to 20 MB.
+Minifying is the process of removing new-lines and whitespace characters from
+the files. The 20 MB number was collected by removing all new-line and
+whitespace characters. In reality, we will need to do something slightly more
+sophisticated and keep whitespace in documentation comments, so that
+type-checking in IDEs still has readable type references.
+
+### Is this a breaking change?
+
+Yes, see CHANGELOG and CLI Notice sections. The solution for the Lambda Layer
+zip files will have a breaking change for customers executing CDK applications
+that use the Lambda Layers in some network-restricted environments. Most
+alternative solutions will also have a breaking change for these customers. The
+proposed solution will have the easiest migration path for these customers.
+
+### What alternative solutions did you consider?
+
+Each category of files has different alternative solutions to reduce the size.
+
+#### .jsii Files
+
+1. The .jsii.tabl.json  and .jsii files are both “pretty-printed.” We can reduce
+   their size to 42 MB (-23%) and 30 MB (-33%) by removing all the whitespace
+   and newlines. We will not take this solution, because the space savings is
+   not as significant as compression. And, then we can maintain the benefit of
+   using pretty-printed JSON when uncompressed.
```

Comment:
I would argue the benefit of pretty-printed JSON isn't that large, and depending on the JSON decoded/parser loading the files would be faster if we strip whitespace. Might not be a big impact, but probably won't hurt either.

I would be in favor of this I think.

Reply:
So humans never look at these files? In that case, I agree, we should stop pretty-printing. Currently thinking in addition to the compression.